From 34c14824642c2cd311252294bfd874e62a6a1d2c Mon Sep 17 00:00:00 2001
From: Aisheng.Dong <b29396@freescale.com>
Date: Thu, 2 Sep 2010 14:41:11 +0800
Subject: [PATCH] ENGR00131466-6 mx28: add spi support

Add spi device, pad setting and set the correct clock to use.

Signed-off-by: Aisheng.Dong <b29396@freescale.com>
---
 arch/arm/mach-mx28/device.c       |   44 +++++++++++++++++++++++++++++++++++++
 arch/arm/mach-mx28/mx28evk.c      |   11 +++++++++
 arch/arm/mach-mx28/mx28evk_pins.c |   43 ++++++++++++++++++++++++++++++++++++
 3 files changed, 98 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx28/device.c b/arch/arm/mach-mx28/device.c
index 068d178..586bdf1 100644
--- a/arch/arm/mach-mx28/device.c
+++ b/arch/arm/mach-mx28/device.c
@@ -659,6 +659,49 @@ static void mx28_init_mmc(void)
 }
 #endif
 
+#if defined(CONFIG_SPI_MXS) || defined(CONFIG_SPI_MXS_MODULE)
+static struct mxs_spi_platform_data spi_data = {
+	.clk = "ssp.2",
+};
+static struct resource ssp2_resources[] = {
+	{
+		.start	= SSP2_PHYS_ADDR,
+		.end	= SSP2_PHYS_ADDR + 0x2000 - 1,
+		.flags	= IORESOURCE_MEM,
+	}, {
+		.start	= MXS_DMA_CHANNEL_AHB_APBH_SSP2,
+		.end	= MXS_DMA_CHANNEL_AHB_APBH_SSP2,
+		.flags	= IORESOURCE_DMA,
+	}, {
+		.start	= IRQ_SSP2_DMA,
+		.end	= IRQ_SSP2_DMA,
+		.flags	= IORESOURCE_IRQ,
+	}, {
+		.start	= IRQ_SSP2,
+		.end	= IRQ_SSP2,
+		.flags	= IORESOURCE_IRQ,
+	},
+};
+
+static void __init mx28_init_spi(void)
+{
+	struct platform_device *pdev;
+
+	pdev = mxs_get_device("mxs-spi", 0);
+	if (pdev == NULL || IS_ERR(pdev))
+		return;
+	pdev->resource = ssp2_resources;
+	pdev->num_resources = ARRAY_SIZE(ssp2_resources);
+	pdev->dev.platform_data = &spi_data;
+
+	mxs_add_device(pdev, 3);
+}
+#else
+static void mx28_init_spi(void)
+{
+}
+#endif
+
 #if defined(CONFIG_MXS_WATCHDOG) || defined(CONFIG_MXS_WATCHDOG_MODULE)
 static struct resource mx28_wdt_res = {
 	.flags = IORESOURCE_MEM,
@@ -1513,6 +1556,7 @@ int __init mx28_device_init(void)
 	mx28_init_lradc();
 	mx28_init_auart();
 	mx28_init_mmc();
+	mx28_init_spi();
 	mx28_init_gpmi_nfc();
 	mx28_init_wdt();
 	mx28_init_rtc();
diff --git a/arch/arm/mach-mx28/mx28evk.c b/arch/arm/mach-mx28/mx28evk.c
index 37beb27..3a4741a 100644
--- a/arch/arm/mach-mx28/mx28evk.c
+++ b/arch/arm/mach-mx28/mx28evk.c
@@ -32,6 +32,8 @@
 #include <mach/device.h>
 #include <mach/pinctrl.h>
 
+#include <linux/spi/spi.h>
+
 #include "device.h"
 #include "mx28evk.h"
 
@@ -44,6 +46,14 @@ static void __init i2c_device_init(void)
 	i2c_register_board_info(0, mxs_i2c_device, ARRAY_SIZE(mxs_i2c_device));
 }
 
+static struct spi_board_info spi_board_info[] __initdata = {
+};
+
+static void spi_device_init(void)
+{
+	spi_register_board_info(spi_board_info, ARRAY_SIZE(spi_board_info));
+}
+
 static void __init fixup_board(struct machine_desc *desc, struct tag *tags,
 			       char **cmdline, struct meminfo *mi)
 {
@@ -97,6 +107,7 @@ static void __init mx28evk_device_init(void)
 {
 	/* Add mx28evk special code */
 	i2c_device_init();
+	spi_device_init();
 	mx28evk_init_leds();
 }
 
diff --git a/arch/arm/mach-mx28/mx28evk_pins.c b/arch/arm/mach-mx28/mx28evk_pins.c
index 7d5b643..3b6afe4 100644
--- a/arch/arm/mach-mx28/mx28evk_pins.c
+++ b/arch/arm/mach-mx28/mx28evk_pins.c
@@ -1072,6 +1072,43 @@ static struct pin_desc mx28evk_gpmi_pins[] = {
 	 },
 };
 
+#if defined(CONFIG_SPI_MXS) || defined(CONFIG_SPI_MXS_MODULE)
+static struct pin_desc mx28evk_spi_pins[] = {
+	{
+	 .name	= "SSP2 MOSI",
+	 .id	= PINID_SSP2_MOSI,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_4MA,
+	 .voltage	= PAD_3_3V,
+	 .drive 	= 1,
+	 },
+	{
+	 .name	= "SSP2 MISO",
+	 .id	= PINID_SSP2_MISO,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_4MA,
+	 .voltage	= PAD_3_3V,
+	 .drive 	= 1,
+	 },
+	{
+	 .name	= "SSP2 SCK",
+	 .id	= PINID_SSP2_SCK,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_4MA,
+	 .voltage	= PAD_3_3V,
+	 .drive 	= 1,
+	 },
+	{
+	 .name	= "SSP2 SS0",
+	 .id	= PINID_SSP2_SS0,
+	 .fun	= PIN_FUN1,
+	 .strength	= PAD_8MA,
+	 .voltage	= PAD_3_3V,
+	 .drive 	= 1,
+	 },
+};
+#endif
+
 #if defined(CONFIG_FEC) || defined(CONFIG_FEC_MODULE)\
 	|| defined(CONFIG_FEC_L2SWITCH)
 int mx28evk_enet_gpio_init(void)
@@ -1174,6 +1211,12 @@ void __init mx28evk_pins_init(void)
 		mx28evk_init_pin_group(mx28evk_gpmi_pins,
 						ARRAY_SIZE(mx28evk_gpmi_pins));
 	}
+
+#if defined(CONFIG_SPI_MXS) || defined(CONFIG_SPI_MXS_MODULE)
+	mx28evk_init_pin_group(mx28evk_spi_pins,
+					ARRAY_SIZE(mx28evk_spi_pins));
+#endif
+
 #if defined(CONFIG_FEC) || defined(CONFIG_FEC_MODULE)\
 	|| defined(CONFIG_FEC_L2SWITCH)
 		mx28evk_init_pin_group(mx28evk_eth_pins,
-- 
1.5.4.4

